packages/truffle-core/lib/command.js

```diff
@@ -1,138 +1,122 @@
+    // for inferring the command.
+    if (this.commands[firstInputString]) {
+      chosenCommand = firstInputString;
+    } else if (noAliases !== true) {
+      let currentLength = 1;
+      const availableCommandNames = Object.keys(this.commands);
+
+      // Loop through each letter of the input until we find a command
+      // that uniquely matches.
+      while (currentLength <= firstInputString.length) {
+        // Gather all possible commands that match with the current length
+        const possibleCommands = availableCommandNames.filter(
+          possibleCommand => {
+            return (
+              possibleCommand.substring(0, currentLength) ===
+              firstInputString.substring(0, currentLength)
+            );
+          }
         );
-      });
 
-      // Did we find only one command that matches? If so, use that one.
-      if (possibleCommands.length === 1) {
-        chosenCommand = possibleCommands[0];
-        break;
-      }
+        // Did we find only one command that matches? If so, use that one.
+        if (possibleCommands.length === 1) {
+          chosenCommand = possibleCommands[0];
+          break;
+        }
 
-      currentLength += 1;
+        currentLength += 1;
+      }
     }
-  }
 
-  if (chosenCommand == null) {
-    return null;
-  }
-
-  var command = this.commands[chosenCommand];
+    if (chosenCommand == null) {
+      return null;
+    }
 
-  return {
-    name: chosenCommand,
-    argv: argv,
-    command: command
-  };
-};
+    const command = this.commands[chosenCommand];
 
-Command.prototype.run = function(inputStrings, options, callback) {
-  if (typeof options === "function") {
-    callback = options;
-    options = {};
+    return {
+      name: chosenCommand,
+      argv,
+      command
+    };
   }
 
-  const result = this.getCommand(inputStrings, options.noAliases);
+  run(inputStrings, options, callback) {
+    const result = this.getCommand(inputStrings, options.noAliases);
 
-  if (result == null) {
-    return callback(
-      new TaskError(
-        "Cannot find command based on input: " + JSON.stringify(inputStrings)
-      )
-    );
-  }
+    if (result == null) {
+      return callback(
+        new TaskError(
+          "Cannot find command based on input: " + JSON.stringify(inputStrings)
+        )
+      );
+    }
 
-  var argv = result.argv;
+    const argv = result.argv;
 
-  // Remove the task name itself.
-  if (argv._) {
-    argv._.shift();
-  }
+    // Remove the task name itself.
+    if (argv._) argv._.shift();
+
+    // We don't need this.
+    delete argv["$0"];
 
-  // We don't need this.
-  delete argv["$0"];
+    const newOptions = Object.assign({}, options, argv);
 
-  // Some options might throw if options is a Config object. If so, let's ignore those options.
-  var clone = {};
-  Object.keys(options).forEach(function(key) {
     try {
-      clone[key] = options[key];
-    } catch (e) {
-      // Do nothing with values that throw.
```

Comment:
Ah so this `try`/`catch` is important because of the way truffle-config works, since it throws exceptions when you try to assign to things that aren't kosher. Might not be safe to remove this.

Reply:
What's confusing is that it seems like `truffle-config` does catch and throw exceptions for non-kosher assignments, but here the code catches and then does nothing? Does the error bubble up in the `catch` on line 119?

If that's the case, then I think it does make sense to keep the `try/catch`, since I believe users will see a confusing `TypeError` thrown instead.

Reply:
@CruzMolina It is attempting to "sanitize" the values. In other words, it will remove all the values that `throw`. I'm not certain that it is necessary though...

Reply:
Ah, I see. `truffle-config` has it's own normalizing method (https://github.com/trufflesuite/truffle/blob/truffle-core-housekeeping/packages/truffle-config/index.js#L297-L307).

I think the comment `Some options might throw...` suggests if options is a `TruffleConfig` object containing getters, the key assignment might throw. I'm not sure if `Object.assign` will appropriately skip a getter and _not_ throw a `TypeError`.

Reply:
@CruzMolina I don't think that `options` will ever be an instance of a config object. Can you think of a case where it would be?

Reply:
Looks like `options` includes an instance or two of a `Config` object while inside `truffle develop` and `truffle console`.

The original comment isn't very specific (would be nice to know _which_ options might throw), but I imagine it would be safer to keep a `try/catch` here to make sure we don't inadvertently break something obscure.

Reply:
Yeah but keep in mind that this is in `command.js` and before any of the code in the "commands" files themselves. `options` at this point will always just be the logger I think. What gets attached later should have nothing to do with this code.

Reply:
Ahh yes I see where a command is instantiated with a config object, you are right. I wonder if it really causes any problems...